man/man1/v6mail.1
.TH V6MAIL 1 "7 Nov 84"
.SH NAME
v6mail \- send mail to designated user(s)
.SH SYNOPSIS
.B v6mail 
[-yn ] 
.br
.B v6mail 
to-addressees ... [\-c cc-addressees ...] [\-f from] [\-g signature] [\-i] [\-r] 
.br
	[\-s subject-text] [\-t more-to-addressees ...] [--submit-args]
.br
	[\-a replyto-text]
.sp
(Sometimes installed as "mail" instead of "v6mail")
.SH DESCRIPTION
This document describes a program which is intended to
be upwards compatible with the original Unix Version 6
.I mail
command
from Bell Labs.  This program uses the MMDF system to perform
distribution.  Differences that are visible to the user lie in
the treatment of individual messages during the display
function, the creation of messages which conform with Arpanet
message format (syntax), and the availability of
additional switches during message creation.

The first form of the
.I v6mail
command (no arguments or exactly one which is a switch "\-y" or
"\-n") shows you any new mail which is in your
MMDF mailbox.  The original Bell Labs
.I mail
command lists the contents of the entire mailbox and
then asks if you want it all saved.  If you indicate yes,
then it is added (pre-pended) to the file "mbox" in your
login directory. The
.I v6mail
command behaves similarly,
except that it queries you about the handling of
.I each
message, rather than all of them together, and the messages
are appended to the
.I end
of mbox.  If a switch "\-y" or "\-n" is
specified, then it is taken to be the response to the "Save?"
queries.  To delete a message, you must answer the query
with a wording starting with "n".  Any other response (including
an emply line) will cause the message to be saved in mbox.
Aborting the command will leave your mailbox and mbox untouched.

The second form of the
.I v6mail
command creates and sends a
message to the indicated recipient(s).  Headers for the
message are created automatically and, optionally, as
indicated by certain command arguments as explained below.
These headers are created in the order implied by the
parameter list.  (Note that this can cause multiple 
header-components of the
same name to be created, separated by other components.)  A
"Date:" component is always created, containing the creation
date and time of the message.  The text body of the message
is read from standard input.  The message which is sent
conforms to current Arpanet standards and can be manipulated by
systems such as \fImsg\fR and MH.  
No draft message is created; so the message may
not be edited after initial creation and before it is sent.

When parameters are not preceded by a switch parameter (i.e.,
one with a dash as its first character) they are taken to be
primary addressees for the mail.  These addressees are then
listed in the message's "To:" component(s).  

A recipient address may consist of the login name of a local
user (e.g., "Joe" or "Smith"), an alias which is
contained in the aliases file (e.g., Project-Leaders) or a full
network name (e.g., "smith@a.cs.okstate.edu").
Obviously, use of network addresses requires the
availability of a network, although reference to the local
host (the one you are using) is always allowed.

.I V6mail
looks for an optional file called ".signature" in the user's home directory
and fills in the signature in the "From:" or "Sender:" component that 
corresponds to the user.  (Make sure that ".signature" contains
one line and that the contents are in double quotes if they contain
non-alphanumeric characters.)  

The "\-t" switch causes following parameters to (again) be
used as addressees in the "To:" header.  Normally, therefore,
the switch is not needed.

The "\-c" switch causes following parameters to be used as
addressees and listed in a "cc:" (carbon copy) header.  This
is in effect until the next switch parameter is encountered.

The "\-s" switch causes the next single parameter to be used
as the text of a "Subject:" component of the message.  Note
that to include more than one word as text, the text must be
surrounded by quotation marks.

The "\-a" switch causes the next single parameter to be used
as the text of a "Reply-to:" component of the message.  Note
that to include more than one word as text, the text must be
surrounded by quotation marks.

The "\-g" switch causes the next single parameter to be used
as the text of the signature in the "From:" component of
the message.  Note that to include more than one word as text, 
the text must be surrounded by quotation marks.
Only the first occurence of either "\-f" or "\-g" is used.  

The "\-f" switch causes the next single parameter to be used
as the text of a "From:" component of the message.  If this
is used, then a "Sender:" component will automatically be
created and contain the address of the user who is actually
running the 
.I v6mail
program.  This is necessary to guarantee
authentication of mail.  This switch is useful when an agent,
such as a daemon process or a secretary, is logged in under
your name and wants to indicate that the agent, rather than you,
issued the note.  Only the first occurence of either "\-f" or "\-g"
is used.

The "\-i" switch instructs 
.I v6mail
to include a "Message-Id:"
component, containing a unique message identification string.
For current usage, this is not normally needed.

The "\-r" switch instructs
.I v6mail
to ignore any problems with delivery.  Normally, the "\-r" switch
is omitted and the sender is notified of any delays or failures
in the delivery process.

The "--" switch can be used to pass arguments on to MMDF's \fIsubmit\fR
program which is responsible for actually processing
(sending) mail.  Only one "--" switch setting may be used and
it may contain several values to be passed to \fIsubmit\fR.  If
more than one "--" switch is specified, the last one will be
used.  It is generally unlikely that you will need to use
this feature; however, two switches might prove useful: "--y"
causes the message to be sent to the user's terminal, if the
user is logged in and the user's screen is writable.  "--w" lets
you watch the delivery process, allowing you to verify that
delivery actually occurs.  This switch, therefore, is useful
when the mail is important and you are not adequately certain
that the mail system can be trusted, such as when it is newly
installed on the system.  To use both switches, simply
specify "--yw".
.SH EXAMPLES
.nf


"v6mail Farber" gives a message with the following headers:

.RS
Date: Sun, 24 Feb 84 12:43:11 EST
From: Dcrocker@UDEE
To: Farber@UDEE
.RE


"v6mail whittake farber \-c dcrocker" gives:

.RS
Date: Sun, 24 Feb 84 12:43:11 EST
From: Dcrocker@UDEE
To: whittake@UDEE, farber@UDEE
cc: dcrocker@UDEE
.RE


"v6mail farber \-c dcrocker \-t sincos@xerox.com \-g 'Dave Crocker' \-s 'test message'" gives:

.RS
Date: Sun, 24 Feb 84 12:43:11 EST
From: Dave Crocker <Dcrocker@UDEE>
Subject: test message
To: Farber@UDEE
To: sincos@xerox.com
cc: dcrocker@UDEE
.RE
.fi
.SH "SEE ALSO"
msg(1), resend(1), send(1), snd(1), submit(8)
.SH BUGS
.I V6mail
cannot be run on a mailbox that is being read with
\fImsg\fR(1).  It will print an error message and exit if a binary
file exists for a mailbox to prevent trashing the mailbox later
with \fImsg\fR.
.I V6mail
should be rewritten to maintain the binary file.
.PP
.I V6mail
doesn't like empty .signature files.


Changes
Changes beta 3
	* Small bug in ap_lex.c hip@a.cs.okstate.edu
	* jromine@ics.uci.edu some changes for Solaris 2.5, Sun Pro
	  4.2 Compiler
	* linux 2.2.7, glibc2.1, egcs 2.91 (redhat 6.0)
	* sco (first attempt)
	* jromine@ics.uci.edu, boundary fix smtpsrvr
	* Small bug in submit/lnk_getaddr.c krueger@mathematik.uni-kl.de

Wed Dec 31 13:14:34 MET 1997	Kai Krueger <krueger@mathematik.uni-kl.de>

	* create file h/config.h.in and moved every define to it.

	* DOASSIGN renamed to HAVE_ASSIGN. Autodetection by configure

	* SECURETTY renamed to HAVE_SECURETTY, Autodetection by configure

	* NODIAL renamed to HAVE_DIAL. default is 'no'
	  may be enanble by configure --enable-dial
	  if compiling phone/smphone channel it wil be turned on by configure.

	* NAMESERVER renamed to HAVE_NAMESERVER. Autodetection by configure.
	  May be overwriten by --disable-nameserver
          - tb_ns.4.2.c and tb_ns.fake.c merged into one file tb_ns.c and
	  seperated by #ifdef HAVE_NAMESERVER...
	  - removed TB_NS in Makefile.com	  

	* HAVE_TCP_WRAPPER renamed to HAVE_LIBWRAP. Autodetection by configure.
	
	* <time.h> included in config.h. It's location is determined by
	  configure

	* determination of <dir.h>, <dirent.h>, a.s.o is done by configure
	  definition an include has moved to config.h

	* determination of index/rindex done by configure and moved to config.h

	* definition of sigtype renamed to RETSIGTYPE. Detection done by
	  configure.

	* return-type of sprintf determined by configure. Define BSD_SPRINTF
	  if char * is returned.

	* uip/Makefile, src/Makefile, lib/Makefile Makefile.com and Makefile
	  are generated by configure.

	* Value of CH_TB in Makefile.com set by configure. If you want
	  sequential tables run configure with the argument --disable-dbm

	* Added define HAVE_SGTTY_H for the existens of <sgtty.h>

	* Added check for <varargs.h> as define HAVE_VARARGS_H.
	  Added check for variable number of arguments if <varargs.h> don't
	  exists. If Compiler don't supports variable number of arguments
	  NO_VARARGS is defined. (NO_VARARGS should be renamed)
	  configure ist checking for both of it. But if <varagrs.h> isn't
	  found lib/util/multcat.c and lib/util/multcpy.c could't be compiled.
	  
	* test configure on machine that dosn't support multiple arguments.
	  (check of NO_VARARGS)

	* test for dup2() and fcntl(fd1, F_DUPFD, fd2) done.
	  NODUP2 replaced by HAVE_DUP2 and
	  NOFCTNL replaced by HAVE_FCNTL_F_DUPFD


uip/other/v6mail.c
#include "util.h"                 /* to get mmdf reply codes            */
#include "mmdf.h"                 /* to get mmdf reply codes            */
#include "cnvtdate.h"
#include <pwd.h>
#include <grp.h>
#include <signal.h>
#include <sys/stat.h>
#include "mm_io.h"

/*  Upgrade to use MMDF Submit; 10 Nov. 1978; David H. Crocker
 *    1 Dec 78 Dave Crocker:  add mail-reading, approx like old "mail"
 *   15 Dec 78 Dave Crocker:  add sndmsg & memo & "-" interfaces
 *   16 Dec 78 Dave Crocker:  "-" not generate return address
 *   25 Apr 80 Dave Crocker:  fix .mail acquisition, default to save
 *   14 Jun 80 Dave Crocker:  turn tty writing off/on
 *   17 Jul 80 Dave Crocker:  fix -c
 *   24 Sep 80 Dave Crocker:  add -r
 *   12 Oct 80 Dave Crocker:  add -g & signature file
 *   15 Nov 80 Dave Crocker:  conversion to V7 package & use of mm_io
 *    1 Dec 80 Dave Crocker:  check success of ALL system calls
 *   20 Jul 81 Dave Crocker:  fix dobody() to detect end from terminal
 *   16 Jul 82 Dave Crocker:  make addresses have '@' only; no " at "
 *    6 Dec 83 Doug Kingston: remove tty write stuff, this is not the place.
 *   19 May 83 Doug Kingston: fix -r to pass a flag to submit
 */

long int    curnewpos,
	    poshack;              /* keep track of current position     */
				  /*   because ftell isn't on V6        */
FILE *  newfp;                    /* handle on .mail file               */

char   *logdir;
char   *username;
char   *mbxname;                  /* full pathname to mailbox           */
char    *binbox;
char    ttyobuf[BUFSIZ];          /* for buffering ttyoutput            */
char    *sbmtargs = (char *) 0;
char    *subject = (char *) 0;
char    *replyto = (char *) 0;
char    *sender = (char *) 0;
char    *from = (char *) 0;
char    tolist[LINESIZE];
char    cclist[LINESIZE];

/* mmdf globals */

extern int sentprotect;           /* default mailbox protection         */
extern char     *mmdfgroup;
extern char     *mldflfil;        /* default mailbox file               */
extern char     *mldfldir;        /* directory containing mailbox file  */
extern char     *delim1;          /* string delimiting messages         */
extern char     *locname;
extern char     *locdomain;

/* end of mmdf globals */


char    noret;                    /* no return address                  */
char    watchit;                  /* user wants to watch delivery       */
char    idflag;                   /* insert a Message ID header line    */
char    qrychar;                  /* user-specified default save status */

char    usertype;
#define U_MAIL  0                 /* the original mail command          */
#define U_SND   1                 /* sndmsg-type prompting              */
#define U_MEMO  2                 /* memo prompting                     */
#define U_LMEM  3                 /* logged in as memo, for memo        */

#define FRMNONE 0                 /* no From field specified            */
#define FRMTXT  1                 /* From field had text only           */
#define FRMSNDR 2                 /* field had sender info, too         */

char    addrtype;
#define ADDR_TO 0                 /* add addresses to the to list       */
#define ADDR_CC 1                 /* add addresses to the cc list       */

/* **************************  MAIN  ******************************* */

main (argc, argv)
int     argc;
char   *argv[];
{
    setbuf (stdout, ttyobuf);
    mmdf_init (argv[0]);
    pgminit ();

    if (initstr ("snd", argv[0], 3)
	    || strindex ("/snd", argv[0]) != -1
	    || initstr ("snd", &argv[0][1], 3)
	    || strindex ("sndmsg", argv[0]) != -1)
	usertype = U_SND;
    else
	if (strindex ("memo", argv[0]) != -1)
	    usertype = U_MEMO;
	else
	    if (argv[0][0] == '-' && argv[0][1] == '\0')
	    {                     /* "-" => login shell   */
		usertype = U_LMEM;
		signal (SIGHUP, SIG_IGN);
				  /* renable these for exiting  */
		prompt ("Send a memo\n");
	    }
    if (usertype == U_MAIL && (argc == 1 || (argc == 2 && *argv[1] == '-')))
	readmail (argc, argv);    /* only when invoked at "mail"  */
    else
	sendmail (argc, argv);
}
/**/

leave (val)
int     val;
{
    if (newfp)
	lk_fclose(newfp, mbxname, (char *)0, (char *)0);
    exit (val == OK ? 0 : 99);
}

RETSIGTYPE pipsig ()
{
    if (rp_gval (endchild (NOTOK)) == RP_NO)
	prompt ("Problem accessing mail submission program\n");
				  /* only comment if it looks like      */
				  /*  submit died ugly                  */
    leave (NOTOK);
}

pgminit ()
{
    extern struct passwd *getpwuid ();
    extern char *getmailid ();
    struct passwd  *pwdptr;
    int     realid,
	    effecid;

    signal (SIGPIPE, pipsig);    /* catch write to bad pipe            */

    getwho (&realid, &effecid);   /* who am i?                          */

    if ((pwdptr = getpwuid (realid)) == (struct passwd *) NULL)
	err_abrt (RP_PARM, "Unable to locate user's name");

    /*  save login directory   */
    logdir = strdup(pwdptr -> pw_dir);

    if ((username = getmailid(pwdptr -> pw_name)) == NULL)
	err_abrt (RP_PARM, "Unable to locate user's mailid");
}
/**/

readmail (argc, argv)
int     argc;
char   *argv[];
{
    register int    count;
    struct group *grpptr;
    
    /* set group-id to mmdfgroup */
    if ( (grpptr = getgrnam(mmdfgroup)) == NULL) {
      printf("Cannot get groupname '%s'\n", mmdfgroup);
      exit;
    }
    if( setgid (grpptr->gr_gid) == NOTOK ) {
      printf("Cannot set groupid (%d)\n", grpptr->gr_gid);
      exit;
    }

    if (argc == 2)
	qrychar = argv[1][1];

    readinit ();
    for (count = 0; msgstrt () && showmsg () && savmsg (); count++);
    if (count == 0)
	err_abrt (RP_OK, "No new mail");
    endread ();
}

readinit ()
{
    extern char *multcat ();
    int     newfd;
    extern int  errno;

    mbxname = multcat (
		((mldfldir == 0 || isnull(*mldfldir)) ? logdir : mldfldir),
		 "/",
		((mldflfil == 0 || isnull(*mldflfil)) ? username : mldflfil),
		  0);
    binbox = multcat (
		((mldfldir == 0 || isnull(*mldfldir)) ? logdir : mldfldir),
		 "/._",
		((mldflfil == 0 || isnull(*mldflfil)) ? username : mldflfil),
		  0);
    if (access(binbox, 0) == 0) {
	fprintf(stderr, "Your mailbox has a binary box (%s),\n", binbox);
	fprintf(stderr, "probably created by another mail reading program like MSG.\n");
	fprintf(stderr, "You should remove the binary box before running this program\n");
	fprintf(stderr, "again, or use MSG instead to read your mail.\n");
	exit (99);
    }

    if ((newfd = lk_open (mbxname, 0, (char *)0, (char *)0, 5)) == -1)
	err_abrt (RP_OK,
		    (errno == ETXTBSY) ? "Mailbox is busy" : "No new mail");

    newfp = fdopen (newfd, "r");
    poshack = 0l;
}

endread ()
{
    fflush (stdout);

    if (creat (mbxname, sentprotect) == NOTOK)
				  /* truncate the file            */
	err_abrt (RP_FCRT, "Unable to reset %s mailbox", mbxname);
    leave (OK);
}
/**/

msgstrt ()
{
    char    linebuf[LINESIZE];

    FOREVER
    {
	curnewpos = poshack;
	if (fgets (linebuf, sizeof linebuf, newfp) == NULL)
	    return (FALSE);
	if (!strequ (linebuf, delim1))
	    break;
	poshack += strlen (linebuf);
    }
    fseek (newfp, curnewpos, 0);
    return (TRUE);
}

showmsg ()
{
    char    linebuf[LINESIZE];
    register int    lines;

    for (lines = 0; fgets (linebuf, sizeof linebuf - 1, newfp) != NULL;)
    {
	poshack += strlen (linebuf);
	if (strequ (linebuf, delim1))
	    return (TRUE);

	fputs (linebuf, stdout);
	if (lines++ > 4)
	{
	    fflush (stdout);
	    lines = 0;
	}
    }
    return ((ferror (newfp) || ferror (stdout)) ? FALSE : TRUE);
}
/**/

savmsg ()
{
    static  FILE * mboxfp;
    char    linebuf[LINESIZE];

    if (!querysav ())
	return (TRUE);

    if (mboxfp == NULL)
    {
	if ((mboxfp = fopen ("mbox", "a")) == NULL ||
		chmod ("mbox", sentprotect) == NOTOK)
	    err_abrt (RP_FOPN, "Unable to access/create mbox");
    }

    poshack = curnewpos;
    for ( fseek (newfp, curnewpos, 0);
	    fgets (linebuf, sizeof linebuf - 1, newfp) != NULL;
	    fputs (linebuf, mboxfp))
    {
	poshack += strlen (linebuf);
	if (strequ (linebuf, delim1))
	    break;
    }
    fputs (delim1, mboxfp);
    fflush (mboxfp);
    return (ferror (newfp) ? FALSE : TRUE);
}

querysav ()
{
    register int  respchar;

    if (qrychar == 0)
    {
	prompt ("\nSave? ");
	switch (respchar = getchar ())
	{
	    case '\n':
		break;
	    case '\0':
		exit (-1);
	    default:
		while (getchar () != '\n');
	}
    }
    else
	respchar = qrychar;
    if (respchar == 'n' || respchar == 'N')
	return (0);               /* deletion must be explicit            */
    return (1);                   /* default to saving                    */
}
/* *************************  SENDING ****************************** */

static char regargs[] = "rmxto,cc*";
  /* quick NS timeout, return to sender, mail, extract addrs from to & cc  */

sendmail (argc, argv)             /* Send a message                     */
int     argc;
char   *argv[];
{
    extern char *sbargs ();
    int     retval;
    char    linebuf[ADDRSIZE];
    char    hadfrom,
	    hadto,
	    hadcc;

    parseargs(argc, argv);

    if (rp_isbad (mm_init ()) || rp_isbad (mm_sbinit ()))
	err_abrt (RP_MECH, "Unable to submit mail; please report this error");

    strncpy (linebuf, regargs, sizeof(linebuf));    /* standard stuff        */
    if (isstr(sbmtargs))
        strcat (linebuf, sbmtargs);   /* extra args for submit   */

    if (usertype == U_LMEM)
	strcat (linebuf, "u");    /* do not verify the from field       */
    if (noret)
	strcat (linebuf, "q");    /* quiet, do not return on errors  */
    if (watchit)
	strcat (linebuf, "w");    /* user wants to watch the process     */

    if (rp_isbad (mm_winit ((char *) 0, linebuf, (char *) 0)))
	err_abrt (RP_MECH, "problem with submit message initialization");

    dodate ();

    hadfrom = dofrom ();

    dosubject ();

    if (hadfrom == FRMTXT && usertype != U_LMEM)
	dosender ("Sender:  ", (char *) 0, TRUE);
				/* simple address */
    hadto = doto ();
    hadcc = docc ();

    if (!hadto && !hadcc && usertype == U_MAIL)
	err_abrt (RP_PARM, "No addressees specified");

    doreply ();

    doid ();

    if (usertype != U_MAIL)
    {
	doprompt ();              /* get headers          */
	prompt ("Text:\n");
    }
    mm_wtxt ("\n", 1);           /* start the body             */


    dobody ();
    retval = endbody ();

    endchild (OK);

    if (usertype != U_MAIL && rp_isgood (retval))
	prompt ("Message posted\n");

    leave ((rp_isbad (retval)) ? NOTOK : OK);
}
/*  *************  ARGUMENT PARSING FOR SENDMAIL  **************  */

parseargs (argc, argv)               
int     argc;
char   *argv[];
{
    register int    i;

    tolist[0] = cclist[0] = '\0';
    addrtype = ADDR_TO;

    for (i = 1; i < argc; i++)
    {
	switch (argv[i][0])
	{
	default:
	    if (addrtype == ADDR_TO) {
		if (!strlen(tolist))
		    strncpy (tolist, argv[i], sizeof(tolist));
		else {
		    strcat (tolist, ", ");
		    strcat (tolist, argv[i]);
		}
	    } else { /* ADDR_CC */
		if (!strlen(cclist))
		    strncpy (cclist, argv[i], sizeof(cclist));
		else {
		    strcat (cclist, ", ");
		    strcat (cclist, argv[i]);
		}
	    }
	    break;
	case '-':
	    switch (argv[i][1])
	    {
		case '-':
		    sbmtargs = &(argv[i][2]);
		    break;

		case 'r':
		    noret = TRUE;
		    break;

		case 'w':
		    watchit = TRUE;
		    break;

		case 'i':
		    idflag = TRUE;
		    break;

		case 's':
		    subject = argv[++i];
		    break;

		case 'a':
		    replyto = argv[++i];
		    break;

		case 'f':
		    if ((++i < argc) && !isstr(sender))
     		         from = argv[i];
		    break;

		case 'g':
		    if ((++i < argc) && !isstr(from))
		        sender = argv[i];
		    break;

		case 't':
		    addrtype = ADDR_TO;
		    break;

		case 'c':
		    addrtype = ADDR_CC;    
		    break;
	    }
	    break;
	}
    }
}

sndhdr (name, contents)
char	*name,
	*contents;
{
    char    linebuf[LINESIZE];

    snprintf (linebuf, sizeof(linebuf), "%-10s%s\n", name, contents);
    mm_wtxt (linebuf, strlen (linebuf));
}
/**/

dodate ()
{
    char    datbuf[64];

    cnvtdate (TIMREG, datbuf);    /* rfc733 format date                 */
    sndhdr ("Date:  ", datbuf);
}

doid ()
{
    char    linebuf[LINESIZE];
    char    datbuf[64];

    if (idflag)
	{
	    cnvtdate (TIMCOM, datbuf);
	    snprintf (linebuf, sizeof(linebuf), "<%s.%d@%s.%s>", datbuf, getpid(),
			locname, locdomain);
	    sndhdr ("Message-Id:  ", linebuf);
	}
}

dosubject ()
{
    if (isstr(subject))
	sndhdr ("Subject:  ", subject);
}

doreply ()
{
    if (isstr(replyto))
	sndhdr ("Reply-To: ", replyto);
}
/**/

dofrom ()
{
    static char fldnam[] = "From:  ";
    register int    i;

    if (isstr(sender))
	{
	    dosender (fldnam, sender, TRUE);
	    return (FRMSNDR);
	}
    else if (isstr(from))
	{
	    sndhdr (fldnam, from);
	    return (FRMTXT);
	}
    else if (usertype != U_LMEM)       /* not memo                     */
	dosender (fldnam, (char *) 0, TRUE);

    return (FRMNONE);
}
/**/

dosender (cmpnt, name, fancy)
char    cmpnt[],
	name[];
int     fancy;          /* make address fancy? */
{
    FILE   *sigfp;
    char   *ptr,
	    linebuf[ADDRSIZE],
	    sigtxt[FILNSIZE];     /* where is signature text?           */
    char    gotsig;

    gotsig = FALSE;

    if (!fancy || name != 0)
    {
	strncpy (sigtxt, name, sizeof(sigtxt));
	gotsig = TRUE;
    }
    else
    {                             /* user didn't give us a signature    */
	snprintf (linebuf, sizeof(linebuf), "%s/.signature", logdir);

	if ((sigfp = fopen (linebuf, "r")) != NULL)
	{
	    if (fgets (sigtxt, sizeof sigtxt, sigfp) != NULL) {
		char *cp;

		if (cp = strchr(sigtxt, '\n'))
		    *cp = '\0';
	    	if (sigtxt[0])
		    gotsig = TRUE;
	    }
	}
    }

    if (gotsig)                   /* real name + mailbox                */
    {
	for (ptr = locname; *ptr != 0; ptr++)
	    *ptr = uptolow (*ptr);      /* Screw locname */
	snprintf (linebuf, sizeof(linebuf), "%s <%s@%s.%s>", 
                  sigtxt, username, locname, locdomain);
    }
    else                          /* just the mailbox info              */
	snprintf (linebuf, sizeof(linebuf), "%s@%s.%s", 
                  username, locname, locdomain);

    sndhdr (cmpnt, linebuf);
}
/**/

doto ()
{

    if (strlen(tolist) > 0) {
	sndhdr ("To:  ", tolist);
	return (TRUE);
    } else
        return (FALSE);

}
/**/

docc ()
{
    if (strlen(cclist) > 0) {
	sndhdr ("cc:  ", cclist);
	return (TRUE);
    } else
	return (FALSE);
}
/**/

doprompt ()
{
    if (usertype == U_LMEM)       /* "free" memo          */
	cpyprompt ("From:  ");
    cpyprompt ("To:  ");
    if (usertype == U_SND)        /* sndmsg               */
	cpyprompt ("cc:  ");
    cpyprompt ("Subject:  ");
}

prompt (str)
char   *str;
{
    printf ("%s", str);
    fflush (stdout);
}

cpyprompt (prom)
char    prom[];
{
    prompt (prom);
    copylin (prom);
}

copylin (prelim)
char   *prelim;
{
    char    linebuf[LINESIZE];
    register int    morin;
    register int    c;

    for (morin = TRUE; morin; sndhdr (prelim, linebuf), prelim = "")
    {
	morin = FALSE;
	if (fgets (linebuf, sizeof linebuf, stdin) == NULL)
	    err_abrt (RP_LIO, "Input error");

	if (linebuf[0] == '\n')
	    return;

	c = strlen (linebuf);
	linebuf[c - 1] = '\0';
	if (linebuf[c - 2] == '\\')
	{
	    morin = TRUE;
	    linebuf[c-- - 2] = '\n';
	}
    }
}
/**/

dobody ()
{
    char    buffer[BUFSIZ];
    register int    i;

    while (!feof (stdin) && !ferror (stdin) &&
	    (i = fread (buffer, sizeof (char), sizeof (buffer), stdin)) > 0)
	if (rp_isbad (i = mm_wtxt (buffer, i)))
		err_abrt (i, "Problem writing body");

    if (ferror (stdin))
	err_abrt (RP_FIO, "Problem reading body");
}

endbody ()
{
    struct rp_bufstruct thereply;
    int     len;

    if (rp_isbad (mm_wtend ()))
	err_abrt (RP_MECH, "problem ending submission");

    if (rp_isbad (mm_rrply (&thereply, &len)))
	err_abrt (RP_MECH, "problem getting submission status");

    if (rp_isbad (thereply.rp_val))
	err_abrt (thereply.rp_val, "%s", thereply.rp_line);

    return (thereply.rp_val);
}

doreset ()
{
    exit (NOTOK);
}

/* *********************  UTILITIES  ***************************  */

endchild (type)
int     type;
{
    int retval;


    if (rp_isgood (retval = mm_sbend ()))
	retval = mm_end (type);

    return ((retval == NOTOK) ? RP_FIO : (retval >> 8));
}

/*VARARGS2*/
err_abrt (code, fmt, b, c, d)     /* terminate the process              */
int     code;                     /* a mmdfrply.h termination code      */
char   *fmt, *b, *c, *d;
{
    if (fmt) {
	printf (fmt, b, c, d);
    	putchar ('\n');
    }
    if (code != RP_OK)
	printf ("Message posting aborted\n");
    fflush (stdout);
    endchild (NOTOK);
    leave (code);
}

